src/shared/components/Projects/ProjectForm.tsx

```diff
@@ -0,0 +1,264 @@
+import * as React from 'react';
+import { Form, Icon, Input, Button, Spin } from 'antd';
+import { WrappedFormUtils } from 'antd/lib/form/Form';
+import { bool } from 'prop-types';
+import { boolean } from '@storybook/addon-knobs';
+
+/**
+ * Custom from controls, based on:
+ * https://ant.design/components/form/#components-form-demo-customized-form-controls
+ */
+export interface PrefixMappingGroupInputState {
+  prefix: string;
+  namespace: string;
+}
+export interface PrefixMappingGroupInputProps {
+  groupId: string | number;
+  value: any;
+  onChange(state: PrefixMappingGroupInputState): void;
+}
+class PrefixMappingGroupInput extends React.Component<
+  PrefixMappingGroupInputProps,
+  PrefixMappingGroupInputState
+> {
+  // Static method called by Form component
+  static getDerivedStateFromProps(nextProps: PrefixMappingGroupInputProps) {
+    if ('value' in nextProps) {
+      return {
+        ...(nextProps.value || {}),
+      };
+    }
+    return null;
+  }
+
+  constructor(props: PrefixMappingGroupInputProps) {
+    super(props);
+    const value = props.value || {};
+    this.state = {
+      prefix: value.prefix || '',
+      namespace: value.namespace || '',
+    };
+  }
+
+  handlePrefixChange = (e: React.FormEvent<HTMLInputElement>) => {
+    const prefix = e.currentTarget.value;
+    this.setState({
+      prefix,
+    });
+    this.props.onChange({ ...this.state, prefix });
+  };
+
+  handleNamespaceChange = (e: React.FormEvent<HTMLInputElement>) => {
+    const namespace = e.currentTarget.value;
+    this.setState({
+      namespace,
+    });
+    this.props.onChange({ ...this.state, namespace });
+  };
+
+  render() {
+    return (
+      <Input.Group
+        compact
+        style={{ width: 'calc(100% - 22px)', marginRight: 8 }}
```

Comment:
Where does this 22px comes from?

Reply:
Can it be explicit somewhere? I'm sure we'll forget this info otherwise

Reply:
I've added a wee comment!